### PR TITLE
fix: error check to ensure path id doesn't conflict body id

### PR DESCRIPTION
**Pull request by jinchizhou**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x ] Bugfix
- [ ] New feature provided
- [ ] Improve performance

- Related issues
fix https://github.com/apache/apisix-dashboard/issues/566
___
### Bugfix
- Description
Ensure Path ID doesn't conflict with Body ID
- Added test case for id to be string and int
- refactored code a bit to reuse a variable.
- 
- How to fix?
Check both the IDs if they both exist
___
### New feature or improvement
- Describe the details and related test reports.

### CODE REVIEW

**Review by juzhiyuan — APPROVED** [review]
LGTM from logics

**Review by juzhiyuan — COMMENTED** [review]
Do we have Testcase to cover this?

**Comment by jinchizhou** [comment]
Nope, I'll add some tests.

**Comment by jinchizhou** [comment]
@nic-chen could you please take a look thanks!

**Comment by jinchizhou** [comment]
FYI, I'm unable to request reviews

File: api/internal/handler/route/route_test.go [hunk 1]
```diff
@@ -754,6 +754,20 @@ func TestRoute(t *testing.T) {
...
+	errRoute := &UpdateInput{}
+	errRoute.ID = "2"
+	err = json.Unmarshal([]byte(reqBody), errRoute)
+	assert.Nil(t, err)
+	ctx.SetInput(errRoute)
+	ret, err = handler.Update(ctx)
+	assert.NotNil(t, err)
+	assert.EqualError(t, err, "ID on path (2) doesn't match ID on body (1)")
+	assert.Equal(t, http.StatusBadRequest, ret.(*data.SpecCodeResponse).StatusCode)
+
```
Comment (nic-chen) [inline api/internal/handler/route/route_test.go:770]: For historical reasons, id supports int and stringing.
so we need test case for int ID in body.

Comment (jinchizhou) [reply api/internal/handler/route/route_test.go:771]: I think this test should support

File: api/internal/handler/route/route_test.go [hunk 2]
```diff
@@ -408,352 +408,380 @@ func TestRoute(t *testing.T) {
...
-	          },
-	          "scale":0.832,
-	          "selected":{
-	              "id":"b93d622c-92ef-48b4-b6bb-57e1ce893ee3",
-	              "type":"node"
-	          }
-	      }
-	  }
-	}`
+	  ` + reqBodyCopy
```
Comment (nic-chen) [inline api/internal/handler/route/route_test.go:750]: we could use a simple route config
```go
reqBody = `{
    "id": "1",
    "uri": "/index.html",
    "upstream": {
        "type": "roundrobin",
        "nodes": [{
            "host": "www.a.com",
            "port": 80,
            "weight": 1
        }]
    }
}`
```

File: api/internal/handler/route/route_test.go [hunk 3]
```diff
@@ -408,352 +408,380 @@ func TestRoute(t *testing.T) {
...
+	errRoute2 := &UpdateInput{}
+	errRoute2.ID = "2"
+	err = json.Unmarshal([]byte(reqBodyErr), errRoute2)
+	assert.Nil(t, err)
+	ctx.SetInput(errRoute2)
+	ret, err = handler.Update(ctx)
+	assert.NotNil(t, err)
+	assert.EqualError(t, err, "ID on path (2) doesn't match ID on body (%!s(float64=1))")
+	assert.Equal(t, http.StatusBadRequest, ret.(*data.SpecCodeResponse).StatusCode)
+
```
Comment (nic-chen) [inline api/internal/handler/route/route_test.go:784]: we should have a test case for IDs are same but ID in body is int:
```sh
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -i -d '
{
    "id": 1,
    "uri": "/index.html",
    "upstream": {
        "type": "roundrobin",
        "nodes": [{
			"host": "www.a.com",
			"port": 80,
			"weight": 1
		}]
    }
}'
```

**Comment by jinchizhou** [comment]
Not sure why these CI tests are blocking. I verified that these tests are failing locally even without my changes.

**Comment by nic-chen** [comment]
> Not sure why these CI tests are blocking. I verified that these tests are failing locally even without my changes.

I think it's due to https://github.com/apache/apisix-dashboard/pull/1067/files#diff-c45dc0115f67204842506d9284c9142195bbca6e06f449fe9fa367c7652e8c74R354-R358

We should judge whether it is a string, and convert it to a string if it's not.

**Comment by jinchizhou** [comment]
I am only converting to a string if the type is float. The body id are guarantee to be a string or a float I assume? I use a new variable so that shouldn't be breaking any existing test cases.

**Comment by membphis** [comment]
@nic-chen need your help ^_^

**Review by nic-chen — APPROVED** [review]

**Comment by nic-chen** [comment]
merged. thanks @jinchizhou

**Pull request merged**